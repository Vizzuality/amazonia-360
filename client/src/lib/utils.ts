import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export const getKeys = Object.keys as <T extends object>(
  obj: T,
) => Array<keyof T>;

export function getContrastColor(hexColor: string): number {
  // Convert hex color to RGB
  const r: number = parseInt(hexColor.substring(1, 3), 16);
  const g: number = parseInt(hexColor.substring(3, 5), 16);
  const b: number = parseInt(hexColor.substring(5, 7), 16);

  // Calculate luminance (brightness) of the color
  const luminance: number = (0.299 * r + 0.587 * g + 0.114 * b) / 255;

  // Return black or white depending on luminance
  return luminance;
}

export function joinWithAnd(arr: string[]) {
  if (arr.length === 0) return "";
  if (arr.length === 1) return arr[0];
  if (arr.length === 2) return arr.join(" and ");

  const lastItem = arr.pop();
  return arr.join(", ") + " and " + lastItem;
}
