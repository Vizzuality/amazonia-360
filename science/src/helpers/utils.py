import json

import geopandas as gpd
from shapely.geometry import Polygon


def transform_polygon_to_rings(geojson):
    """
    Transform a GeoJSON polygon into a format compatible with the ArcGIS API.

    Parameters:
    geojson (dict): A GeoJSON dictionary.

    Returns:
    str: A JSON string with the polygon coordinates.
    """
    poly = Polygon(geojson["geometry"]["coordinates"][0])
    gdf = gpd.GeoDataFrame(index=[0], crs="epsg:4326", geometry=[poly])
    gdf = gdf.to_crs(epsg=102100)
    poly = gdf.geometry[0]
    coordinates = list(poly.exterior.coords)
    # Transform coordinates into the required format
    rings = [[list(coord) for coord in coordinates]]
    poly_dict = {"rings": rings}
    poly_json = json.dumps(poly_dict)
    return poly_json
