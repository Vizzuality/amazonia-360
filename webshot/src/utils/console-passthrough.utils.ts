import { ConsoleMessage } from "playwright";
import { inspect } from "util";

export const consolePassthrough = (msg: ConsoleMessage) => {
  for (let i = 0; i < msg.args().length; ++i) {
    msg
      .args()
      [i].jsonValue()
      .then((value) => console.debug(inspect(value, undefined, 5)))
      .catch((err) => console.debug(`Failed to get value: ${err}`));
  }
};
