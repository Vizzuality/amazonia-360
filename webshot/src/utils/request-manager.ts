import { Logger } from "@nestjs/common";
import { debounce } from "lodash";
import { Page } from "playwright";

export const createRequestManager = (
  page: Page,
  logger: Logger,
  timeoutMs: number = 30_000,
  debounceDelayMs: number = 10_000,
) => {
  const activeRequests: Set<string> = new Set();
  let timeoutId: NodeJS.Timeout | undefined = undefined;

  let resolve: (() => void) | undefined = undefined;
  let reject: (() => void) | undefined = undefined;
  const promise = new Promise<void>((_resolve, _reject) => {
    resolve = _resolve;
    reject = _reject;
  });

  // This function is debounced so that every time a request is started, finished or failed, a
  // small delay is created before we check again if the network is idle
  const checkIfIdle = debounce(() => {
    if (activeRequests.size === 0) {
      if (timeoutId !== undefined) {
        clearTimeout(timeoutId);
      }
      logger.log("Network is idle");
      resolve?.();
    }
  }, debounceDelayMs);

  return {
    setupEvents: () => {
      page.on("request", (request) => {
        const url = request.url();
        // blob requests are ignored as they may be linked to Mapbox WebWorkers and the front-end
        // application doesn't create any
        if (!url.startsWith("blob:")) {
          activeRequests.add(url);
          checkIfIdle();
        }
      });

      page.on("requestfinished", (request) => {
        const url = request.url();
        activeRequests.delete(url);
        checkIfIdle();
      });

      page.on("requestfailed", (request) => {
        const url = request.url();
        activeRequests.delete(url);
        checkIfIdle();
      });
    },
    isIdle: () => {
      // Reject if the timeout has passed
      timeoutId = setTimeout(() => {
        reject?.();
      }, timeoutMs);

      return promise;
    },
  };
};
