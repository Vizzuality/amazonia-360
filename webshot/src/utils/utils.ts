/**
 * Namespace containing utility functions for value validation.
 */
export class Value {
  /**
   * Checks if a value is neither null nor undefined.
   *
   * @template T - The type of the value being checked
   * @param a - The value to check
   * @returns A type predicate indicating if the value is valid (not null or undefined)
   *
   * @example
   * ```typescript
   * const value: string | null = getSomeValue();
   * if (Value.isValid(value)) {
   *   // value is typed as string here
   *   console.log(value.length);
   * }
   * ```
   */
  static isValid = <T>(a: T | null | undefined): a is T => a !== undefined && a !== null;
}
