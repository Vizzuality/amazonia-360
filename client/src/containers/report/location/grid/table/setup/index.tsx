"use client";

import { ChangeEvent } from "react";
import { useCallback, useState } from "react";

import { LuSettings2 } from "react-icons/lu";

import { useGetGridMeta } from "@/lib/grid";

import { useSyncGridDatasets, useSyncGridFilters, useSyncGridFiltersSetUp } from "@/app/store";

import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Label } from "@/components/ui/label";
import { Popover, PopoverTrigger, PopoverContent } from "@/components/ui/popover";
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components/ui/select";

const RANKING_DIRECTION = [
  {
    key: "asc",
    label: "Bottom",
  },
  {
    key: "desc",
    label: "Top",
  },
];

export default function GridTableSetup() {
  const [isOpen, setIsOpen] = useState(false);

  const [gridDatasets, setGridDatasets] = useSyncGridDatasets();
  const [gridFilters] = useSyncGridFilters();
  const [, setGridFiltersSetUp] = useSyncGridFiltersSetUp();
  const [selectedDirection, setDirection] = useState<"asc" | "desc">("asc");
  const [selectedDataset, setDataset] = useState<string>(gridDatasets[0]);
  const [selectedLimit, setSelectedLimit] = useState<number>();

  const { data: rankingOptions } = useGetGridMeta({
    select: (data) =>
      gridDatasets.map((d) => ({
        key: d,
        label: data?.datasets?.find((dataset) => dataset.var_name === d)?.label,
      })),
  });

  const handleRankingChange = useCallback(
    (e: string) => {
      setDataset(e);
    },
    [setDataset],
  );

  const handleDirectionChange = useCallback((e: "asc" | "desc") => {
    setDirection(e);
  }, []);

  const onInputChange = useCallback(
    (v: ChangeEvent<HTMLInputElement>) => {
      const value = parseFloat(v.target.value);
      setSelectedLimit(value);
    },
    [setSelectedLimit],
  );

  const handleFilters = useCallback(() => {
    if (selectedDataset) {
      const updatedOrder = [selectedDataset, ...gridDatasets.filter((d) => d !== selectedDataset)];
      setGridDatasets(updatedOrder);
    }

    if (selectedDirection) {
      setGridFiltersSetUp((prev) => ({
        ...prev,
        direction: selectedDirection,
      }));
    }

    if (selectedLimit) {
      setGridFiltersSetUp((prev) => ({
        ...prev,
        limit: selectedLimit,
      }));
    }
    setIsOpen(false);
  }, [
    selectedDataset,
    selectedDirection,
    selectedLimit,
    gridDatasets,
    setGridDatasets,
    setGridFiltersSetUp,
  ]);

  return (
    <Popover open={isOpen} onOpenChange={setIsOpen}>
      <PopoverTrigger>
        <LuSettings2 className="h-4 w-4" />
      </PopoverTrigger>
      <PopoverContent
        side="right"
        alignOffset={-26}
        sideOffset={32}
        align="start"
        className="w-auto max-w-80 bg-background p-2"
      >
        <div className="space-y-4 p-4">
          <div className="space-y-1">
            <h2 className="text-sm font-semibold text-popover-foreground">Ranking setup</h2>
            <span className="text-sm text-muted-foreground">Define the order of the cells</span>
          </div>
          <div className="w-full gap-2 overflow-hidden">
            <span className="text-sm text-foreground">Order by</span>
            <div className="flex gap-2">
              <Select value={selectedDataset} onValueChange={handleRankingChange}>
                <SelectTrigger className="h-10 w-full max-w-36 rounded-sm">
                  <SelectValue className="text-sm">
                    <p className="flex-grow truncate md:max-w-80 lg:max-w-none">
                      {rankingOptions?.find((opt) => opt.key === selectedDataset)?.label ||
                        "selectedDataset"}
                    </p>
                  </SelectValue>
                </SelectTrigger>
                <SelectContent className="no-scrollbar max-h-96 overflow-y-auto border-none shadow-md">
                  {rankingOptions &&
                    rankingOptions.map((opt) => (
                      <SelectItem key={opt.key} value={opt.key} className="cursor-pointer">
                        {opt.label}
                      </SelectItem>
                    ))}
                </SelectContent>
              </Select>
              <Select value={selectedDirection} onValueChange={handleDirectionChange}>
                <SelectTrigger className="h-10 w-32 rounded-sm">
                  <SelectValue className="text-sm">
                    <p className="w-fit max-w-64 flex-1 truncate md:max-w-80 lg:max-w-none">
                      {RANKING_DIRECTION.find((opt) => opt.key === selectedDirection)?.label || ""}
                    </p>
                  </SelectValue>
                </SelectTrigger>
                <SelectContent className="no-scrollbar max-h-96 overflow-y-auto border-none shadow-md">
                  {RANKING_DIRECTION &&
                    RANKING_DIRECTION.map((opt) => (
                      <SelectItem key={opt.key} value={opt.key} className="cursor-pointer">
                        {opt.label}
                      </SelectItem>
                    ))}
                </SelectContent>
              </Select>
            </div>
          </div>
          <div className="gap-2">
            <Label htmlFor="cells-number" className="text-sm text-foreground">
              Max number of cells
            </Label>
            <Input
              placeholder={`${gridFilters?.limit?.[0]}`}
              id="cell-number"
              type="number"
              min={0}
              className="h-10"
              onChange={onInputChange}
            />
            <span className="text-sm text-muted-foreground">
              Higher numbers will take more time.
            </span>
          </div>
          <div className="flex justify-end">
            <Button onClick={handleFilters}>Apply</Button>
          </div>
        </div>
      </PopoverContent>
    </Popover>
  );
}
