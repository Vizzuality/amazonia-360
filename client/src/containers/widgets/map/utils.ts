export const FALLBACK_WIDGET_DEFAULT_BASEMAP_ID: BasemapIds = "gray-vector";
import { omit } from "@/lib/utils";

import { Indicator } from "@/app/local-api/indicators/route";
import {
  IndicatorView,
  DefaultTopicConfig,
  DefaultTopicIndicatorConfig,
  TopicView,
} from "@/app/parsers";

import { BasemapIds } from "@/components/map/controls/basemap";

type MapIndicatorProperties = "basemapId" | "opacity";

export const handleMapIndicatorPropertyChange = (
  propertyName: MapIndicatorProperties,
  propertyValue: BasemapIds | number,
  overviewTopicsData: DefaultTopicConfig[] | null,
  indicator: Indicator,
  setSyncDefaultTopics: (
    callback: (prevDefaultTopics: DefaultTopicConfig[] | null) => DefaultTopicConfig[] | null,
  ) => void,
  setTopics: (callback: (prevTopicsState: TopicView[] | null) => TopicView[] | null) => void,
  defaultValues: { basemapId: BasemapIds; opacity: number },
) => {
  const currentIndicatorTopicId = indicator.topic.id;
  const defaultValue = defaultValues[propertyName];
  const isResettingToDefault = propertyValue === defaultValue;

  const isManagingViaSyncDefaultTopics = overviewTopicsData?.some(
    (topic) => currentIndicatorTopicId === topic.id,
  );

  if (isManagingViaSyncDefaultTopics) {
    setSyncDefaultTopics(
      (prevDefaultTopics: DefaultTopicConfig[] | null): DefaultTopicConfig[] | null => {
        const newTopicsArray: DefaultTopicConfig[] = prevDefaultTopics
          ? [...prevDefaultTopics]
          : [];
        const topicIndex = newTopicsArray.findIndex((t) => t.id === currentIndicatorTopicId);

        if (topicIndex !== -1) {
          const targetTopic = { ...newTopicsArray[topicIndex] };
          const indicators: DefaultTopicIndicatorConfig[] = targetTopic.indicators
            ? [...targetTopic.indicators]
            : [];
          const indicatorConfigIndex = indicators.findIndex((i) => i.id === indicator.id);

          if (isResettingToDefault) {
            if (indicatorConfigIndex !== -1) {
              const indicatorConfig = { ...indicators[indicatorConfigIndex] };
              delete indicatorConfig[propertyName];

              // eslint-disable-next-line @typescript-eslint/no-unused-vars
              const { id, ...overrides } = indicatorConfig;
              if (Object.keys(overrides).length > 0) {
                indicators[indicatorConfigIndex] = indicatorConfig;
              } else {
                indicators.splice(indicatorConfigIndex, 1);
              }
            }
          } else {
            if (indicatorConfigIndex !== -1) {
              indicators[indicatorConfigIndex] = {
                ...indicators[indicatorConfigIndex],
                [propertyName]: propertyValue,
              };
            } else {
              indicators.push({
                id: indicator.id,
                [propertyName]: propertyValue,
              } as DefaultTopicIndicatorConfig);
            }
          }

          if (indicators.length === 0) {
            newTopicsArray.splice(topicIndex, 1);
          } else {
            newTopicsArray[topicIndex] = {
              ...targetTopic,
              indicators,
            };
          }
        } else if (!isResettingToDefault) {
          newTopicsArray.push({
            id: currentIndicatorTopicId,
            indicators: [
              {
                id: indicator.id,
                [propertyName]: propertyValue,
              } as DefaultTopicIndicatorConfig,
            ],
          });
        }

        return newTopicsArray.length > 0 ? newTopicsArray : null;
      },
    );
  } else {
    setTopics((prevTopicsState: TopicView[] | null): TopicView[] | null => {
      const currentTopics = prevTopicsState || [];
      const topicIndex = currentTopics.findIndex((t) => t.id === indicator.topic.id);

      if (topicIndex === -1) {
        return prevTopicsState;
      }

      const topicToUpdate = currentTopics[topicIndex];
      const indicatorExists = (topicToUpdate.indicators || []).some(
        (ind: IndicatorView) => ind.id === indicator.id && ind.type === "map",
      );

      if (!indicatorExists) {
        return prevTopicsState;
      }

      const newTopics = currentTopics.map((topic, index) => {
        if (index === topicIndex) {
          return {
            ...topic,
            indicators: (topic.indicators || []).map((ind) => {
              if (ind.id === indicator.id && ind.type === "map") {
                if (isResettingToDefault) {
                  return omit(ind, [propertyName]);
                } else {
                  return { ...ind, [propertyName]: propertyValue };
                }
              }
              return ind;
            }),
          };
        }
        return topic;
      });

      return newTopics.length > 0 ? newTopics : null;
    });
  }
};
