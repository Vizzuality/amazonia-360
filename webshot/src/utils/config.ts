import config from "config";
import { Value } from "./utils";

/**
 * Config class to get configuration values from the config package
 */
export class Config {
  /**
   * Get a configuration value
   *
   * @param key The configuration key
   * @param defaultValue The default value if the key is not found
   * @returns The configuration value or the default value if the key is not found
   *
   * @throws Error if the key is not found and no default value is provided
   */
  public static get<T>(key: string, defaultValue?: T): T {
    if (config.has(key)) {
      const value = config.get<T>(key);

      if (Value.isValid(value)) {
        return value;
      } else if (Value.isValid(defaultValue)) {
        return defaultValue;
      }
    } else if (Value.isValid(defaultValue)) {
      return defaultValue;
    }

    throw new Error(
      `Config key ${key} is not defined and no default value is provided`,
    );
  }

  /**
   * Get a string configuration value
   *
   * @param key The configuration key
   * @param defaultValue The default value if the key is not found
   * @returns The configuration value or the default value if the key is not found
   *
   * @throws Error if the key is not found and no default value is provided
   * @throws Error if the configuration value is not a string
   */
  public static getString(key: string, defaultValue?: string): string {
    if (config.has(key)) {
      const value = config.get<string>(key);

      if (typeof value === "string") {
        return value;
      }

      throw new Error(`Config key ${key} is not a string`);
    } else if (Value.isValid(defaultValue)) {
      return defaultValue;
    }

    throw new Error(
      `Config key ${key} is not defined and no default value is provided`,
    );
  }

  /**
   * Get a boolean configuration value
   *
   * @param key The configuration key
   * @param defaultValue The default value if the key is not found
   * @returns The configuration value or the default value if the key is not found
   *
   * @throws Error if the key is not found and no default value is provided
   * @throws Error if the configuration value is not a boolean
   */
  public static getBoolean(key: string, defaultValue?: boolean): boolean {
    if (config.has(key)) {
      const value = config.get<boolean>(key);

      if (typeof value === "boolean") {
        return value;
      }

      throw new Error(`Config key ${key} is not a boolean`);
    } else if (Value.isValid(defaultValue)) {
      return defaultValue;
    }

    throw new Error(
      `Config key ${key} is not defined and no default value is provided`,
    );
  }

  /**
   * Get a number configuration value
   *
   * @param key The configuration key
   * @param defaultValue The default value if the key is not found
   * @returns The configuration value or the default value if the key is not found
   *
   * @throws Error if the key is not found and no default value is provided
   * @throws Error if the configuration value is not a number
   */
  public static getNumber(key: string, defaultValue?: number): number {
    if (config.has(key)) {
      const value = config.get<number>(key);

      if (typeof value === "number") {
        return value;
      } else if (typeof Number(value) === "number" && !isNaN(Number(value))) {
        return Number(value);
      }

      throw new Error(`Config key ${key} is not a number`);
    } else if (Value.isValid(defaultValue)) {
      return defaultValue;
    }

    throw new Error(
      `Config key ${key} is not defined and no default value is provided`,
    );
  }

  /**
   * Retrieves a configuration value as either a string array or a single string.
   *
   * This method looks for a string value under the provided key and processes it as follows:
   * - If the value contains commas, it splits the string and returns the resulting array.
   * - If the value has no commas, it returns the single string value.
   *
   * @param {string} key - The configuration key to retrieve
   * @param {string[] | string} [defaultValue] - The default value to return if the key isn't found
   * @returns {string[] | string} An array of strings if commas are present, otherwise a single string
   *
   * @throws {Error} If the key doesn't exist and no default value is provided
   * @throws {Error} If the value exists but is not a string
   *
   * @example
   * // If config has "allowedOrigins=http://localhost:3000,https://example.com"
   * const origins = Config.getStringAsArray("allowedOrigins");
   * // returns ["http://localhost:3000", "https://example.com"]
   *
   * @example
   * // If config has "defaultLanguage=en"
   * const language = Config.getStringAsArray("defaultLanguage");
   * // returns "en"
   */
  public static getStringAsArray(
    key: string,
    defaultValue?: string[] | string,
  ): string[] | string {
    if (config.has(key)) {
      const value = config.get<string>(key);

      if (typeof value === "string") {
        const values = value.split(",");
        return values.length > 1 ? values : values[0];
      }

      throw new Error(`Config key ${key} is not a string`);
    } else if (Value.isValid(defaultValue)) {
      return defaultValue;
    }

    throw new Error(
      `Config key ${key} is not defined and no default value is provided`,
    );
  }

  /**
   * Retrieves and merges values from two config sources into a single array.
   *
   * This method looks for values in two places:
   * 1. An array stored under the provided `arrayKey`
   * 2. A comma-separated string stored under the provided `stringKey`
   *
   * It combines all values found in either source into a single array with duplicates removed.
   * If neither source contains values, returns the provided default value.
   *
   * @template T - The type of elements in the array
   * @param {string} arrayKey - The config key to look for an array of values
   * @param {string} stringKey - The config key to look for a comma-separated string of values
   * @param {T[]} [defaultValue=[]] - Default value to return if no values are found
   * @returns {T[]} An array containing all unique values from both sources, or the default value if no values were found
   *
   * @remarks
   * When retrieving values from the string source, each value is cast to type T using `as unknown as T`.
   * This may cause type safety issues if the string values can't be properly converted to type T.
   */
  public static getMergedArray<T>(
    arrayKey: string,
    stringKey: string,
    defaultValue: T[] = [],
  ): T[] {
    const values: Set<T> = new Set<T>();

    if (config.has(arrayKey)) {
      const value = config.get<T[]>(arrayKey);

      if (Array.isArray(value)) {
        value.forEach((value) => {
          values.add(value);
        });
      } else {
        console.warn(`Config ${arrayKey} is not an array`);
      }
    } else {
      console.warn(`Config ${arrayKey} does not exist`);
    }

    if (config.has(stringKey)) {
      const value = config.get<string>(stringKey);

      if (typeof value === "string") {
        value.split(",").forEach((value) => {
          values.add(value as unknown as T);
        });
      } else {
        console.warn(`Config ${stringKey} is not a string`);
      }
    } else {
      console.warn(`Config ${stringKey} does not exist`);
    }

    return values.size > 0 ? Array.from(values) : defaultValue;
  }
}
