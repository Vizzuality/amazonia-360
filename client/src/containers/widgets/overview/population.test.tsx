import WidgetPopulation from "@/containers/widgets/overview/population";

import { mockUseGetRasterAnalysis } from "@/jest.setup";
import { render, screen } from "@/jest.utils";

describe("WidgetPopulation", () => {
  beforeEach(() => {
    mockUseGetRasterAnalysis.mockReturnValue({
      data: 100000,
      isFetching: false,
      isFetched: true,
    });
  });

  test("renders WidgetPopulation component without crashing", async () => {
    const { container } = render(<WidgetPopulation />);
    expect(container).toMatchSnapshot();
    expect(screen.getByRole("banner")).toBeInTheDocument();
    expect(screen.getByText("Population")).toBeInTheDocument();
    expect(screen.getByText("inhabitants")).toBeInTheDocument();
    expect(screen.getByText("100,000")).toBeInTheDocument();
  });

  test("renders WidgetPopulation component with loading state", async () => {
    mockUseGetRasterAnalysis.mockReturnValue({
      data: undefined,
      isFetching: true,
      isFetched: false,
    });

    const { container } = render(<WidgetPopulation />);
    expect(container).toMatchSnapshot();
    expect(screen.getByTestId("card-loader")).toBeInTheDocument();
  });
});
