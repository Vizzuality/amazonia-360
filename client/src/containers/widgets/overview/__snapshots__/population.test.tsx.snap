// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetPopulation renders WidgetPopulation component with loading state 1`] = `
<div>
  <div
    class="flex h-full grow break-inside-avoid flex-col overflow-hidden rounded-2xl border border-blue-100 bg-white p-6"
  >
    <header
      class="flex items-start justify-between"
    >
      <h2
        class="text-base text-blue-600"
      >
        Population
      </h2>
      <button
        aria-controls="radix-:r5:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="flex h-6 w-6 items-center justify-center"
        data-state="closed"
        type="button"
      >
        <svg
          class="text-blue-600"
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="M12 16v-4"
          />
          <path
            d="M12 8h.01"
          />
        </svg>
      </button>
    </header>
    <div
      class="animate-pulse rounded-md bg-primary/10 h-12"
      data-testid="card-loader"
    />
  </div>
</div>
`;

exports[`WidgetPopulation renders WidgetPopulation component without crashing 1`] = `
<div>
  <div
    class="flex h-full grow break-inside-avoid flex-col overflow-hidden rounded-2xl border border-blue-100 bg-white p-6"
  >
    <header
      class="flex items-start justify-between"
    >
      <h2
        class="text-base text-blue-600"
      >
        Population
      </h2>
      <button
        aria-controls="radix-:r1:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="flex h-6 w-6 items-center justify-center"
        data-state="closed"
        type="button"
      >
        <svg
          class="text-blue-600"
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="M12 16v-4"
          />
          <path
            d="M12 8h.01"
          />
        </svg>
      </button>
    </header>
    <div>
      <div
        class="flex items-end space-x-2"
      >
        <span
          class="text-4xl font-bold text-blue-600"
        >
          100,000
        </span>
        <span
          class="relative bottom-1.5 text-xs font-medium text-gray-500"
        >
          inhabitants
        </span>
      </div>
    </div>
  </div>
</div>
`;
